### DISCUSSION TITLE
Injecting Mock JsonWebToken With Qualifiers

### DISCUSSION THREAD

**Discussion by InfoSec812** [quarkusio/quarkus#28532] [Q&A] [answered]
According to the documentation [here](https://quarkus.io/blog/mocking/#using-injectmock) I should be able to do the following:

```
@InjectMock
@IdToken
JsonWebToken idToken;
```

But when I try to run my integration tests I get errors like the following:

```
[1] Unsatisfied dependency for type org.eclipse.microprofile.jwt.JsonWebToken and qualifiers [@IdToken]
        - java member: com.redhat.consultant360.api.AbstractApiController#idToken
        - declared on CLASS bean [types=[com.redhat.consultant360.api.EmployeeApi, com.redhat.consultant360.api.AbstractApiController, com.redhat.consultant360.api.EmployeeApiController, java.lang.Object], qualifiers=[@Default, @Any], target=com.redhat.consultant360.api.EmployeeApiController]
```

Could someone help me to understand the *CORRECT* way to inject a mock with qualifiers?

**Comment by sberyozkin** [comment, marked as answer]
@InfoSec812 Have a look at https://quarkus.io/guides/security-openid-connect#integration-testing-security-annotation

For example, see [this test](https://github.com/quarkusio/quarkus/blob/main/integration-tests/oidc-code-flow/src/test/java/io/quarkus/it/keycloak/TestSecurityLazyAuthTest.java#L29).

**Reply by InfoSec812** [reply]
Thanks @sberyozkin and that definitely **CAN** resolve my issue. I am wondering if there might be another option which could be parameterized? Before switching to Quarkus OIDC were were able to inject JsonWebToken using parameterization to test different access restrictions and using per-method annotations will add a lot of work.

**Comment by sberyozkin** [comment]
@InfoSec812, I see. 
@mkouba, hi Martin, is there a way to have a qualifier like `@IdToken` used  with the parameterized mocks ?

**Reply by InfoSec812** [reply]
Actually, it looks like I should be able to use the [WireMock OIDC server](https://quarkus.io/guides/security-openid-connect#integration-testing-wiremock) to accomplish this, but I am not having much luck so far.

**Reply by mkouba** [reply]
> @InfoSec812, I see. @mkouba, hi Martin, is there a way to have a qualifier like `@IdToken` used with the parameterized mocks ?

Well, the error message is clear that `@IdToken` is considered a required qualifier so it should be taken into account. Does the mocked bean also define this qualifier?

**Comment by sberyozkin** [comment]
@InfoSec812 

>Actually, it looks like I should be able to use the [WireMock OIDC server](https://quarkus.io/guides/security-openid-connect#integration-testing-wiremock) to accomplish this, but I am not having much luck so far.

Wiremock OIDC server is used to replace a connection to the actual OIDC container, it does not provide any producers for JsonWebToken, for the OIDC specific solution use `@TestSecurity` alongside `@OidcSecurity` which is in  `quarkus-test-security-oidc`, as suggested earlier.

Or if I understand Martin's comment correctly, you need to add a `quarkus-oidc` dependency anyway as the `@IdToken` is only available in in `quarkus-oidc`

**Reply by InfoSec812** [reply]
I think that I will probably post a different question with a different context because using the WireMock OIDC Server is a completely different approach. Thanks for the help!
